### PR TITLE
Job Duplication

### PR DESCRIPTION
**Pull request by fruttasecca**
## Description

This PR introduces job duplication, which allows duplicating a job, creating a job with the latest code and environments but inheriting the settings of the job that is being duplicated.

The environment variables of the new job are resolved by merging the current project and pipelines environment variables, then adding the inherited job environment variables. Note that this implies that environment variables that were "inherited" by the job that is being duplicated from project/env variables cannot be distinguished from environment variables that were defined in the job. This might be of interest given that if those project/pipeline variables do not exist anymore it might make more sense to not include them in the newly defined job.

Job parameters are resolved by:
 - removing parameters that do not exist anymore.
 - adding new parameters, i.e. parameters that exist in the latest version of the pipeline but do not exist for the job that is being duplicated.
  - parameters that are not to be removed nor are new, i.e. parameters that exist both for the latest pipeline definition and the job use the values that were defined in the job that is being duplicated.
  - this is true for both the strategy json and the job parameters, where each job parameter effectively represents the parameterization of a run.
  - the job parameters are updated in a way that, from the client POV and given the logic in the job view, preserves the runs selection and their ordering, along with the constraint that every run parameterization should be unique.

If the job that is being duplicated is a cronjob, the new job will have its schedule.

Fixes: #144 (partially) 

### Checklist

- [X] I have manually tested the application to make sure the changes don’t cause any downstream issues, which includes making sure `./orchest status --ext` is not reporting failures when Orchest is running.

### Things that need to be discussed further:
- where in the UI should job duplication happen? It's currently done by a button in the `JobView`, which implies that only non `DRAFT` jobs can be duplicated (although it's technically possible to duplicate `DRAFT` jobs)
- env variables inheritance (see the env variables paragraph)

Current UI changes
image

### CODE REVIEW

**Comment by yannickperrenet** [comment]
First off, I like the decisions you made (and explained in the PR message) around the parameters/env vars to choose on overlap etc. So well done :)

I think we can make an additional improvement where "duplicate job" duplicates the job (with this new implementation) but also renames the job to something like "Duplicate of <name>" (similar to how Google Drive does it).

File: services/orchest-webserver/client/src/views/JobView.tsx [hunk 1]
```diff
@@ -448,6 +512,13 @@ const JobView: React.FC<TViewProps> = (props) => {
             icon="arrow_back"
             onClick={returnToJobs.bind(this)}
           />
+          <span className="float-right">
+            <MDCButtonReact
+              label="Duplicate job"
```
Comment (yannickperrenet) [inline services/orchest-webserver/client/src/views/JobView.tsx:517]: Could the label "Duplicate job" be confusing to the users? Depending on how someone interprets "duplicating" it might or might not do what the person expects. It is good to keep in mind here that a user doesn't necessarily know that the only the job configuration (whatever that may mean) is duplicated, but new code and new environments are used.

File: services/orchest-webserver/app/app/views/orchest_api.py [hunk 2]
```diff
@@ -175,75 +172,101 @@ def catch_api_proxy_jupyter_builds_most_recent():
...
-        json_obj = request.json
+        try:
+            job_spec = jobs.create_job_spec(request.json)
+        except error.EnvironmentsDoNotExist as e:
+            return (
+                jsonify(
+                    {
+                        "message": environments_missing_msg.format(
+                            missing_environment_uuids=[",".join(e.environment_uuids)]
+                        ),
```
Comment (yannickperrenet) [inline services/orchest-webserver/app/app/views/orchest_api.py:195]: Wouldn't it be more helpful if we add the environment names to the message as well?

This would require a change in the `create_job_spec` method.

File: services/orchest-webserver/app/app/core/jobs.py [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+    job_spec["pipeline_name"] = parent_job["pipeline_name"]
+    job_spec["pipeline_run_spec"] = {"run_type": "full", "uuids": []}
+
+    if (
+        Project.query.filter_by(
+            uuid=job_spec["project_uuid"],
+        ).one_or_none()
+        is None
+    ):
+        raise error.ProjectDoesNotExist()
```
Comment (yannickperrenet) [inline services/orchest-webserver/app/app/core/jobs.py:131]: I think it is indeed good to add this case even though with the current implementation the user can't duplicate the job if the containing project was first deleted. :+1:

File: services/orchest-webserver/app/app/core/jobs.py [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+    get_environments,
+    get_environments_from_pipeline_json,
+    get_pipeline_json,
+    get_project_directory,
+    pipeline_uuid_to_path,
+)
+
+
+def create_job_spec(config) -> dict:
+    """Returns a job spec given the provided configuration.
```
Comment (yannickperrenet) [inline services/orchest-webserver/app/app/core/jobs.py:23]: Nitpick: "Returns a job spec based on the provided configuration"

File: services/orchest-webserver/app/app/core/jobs.py [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+    # and adding env variables that were not part of the project and
+    # pipeline env vars. NOTE: we currently have no way to discern
+    # old env variables that the job inherited from project/pipeline
+    # env variables. If we could, old project/pipeline env vars that
+    # do not exist anymore could be discarded.
+    job_spec["env_variables"] = {
+        **project_env_vars,
+        **pipeline_env_vars,
+        **parent_env_vars,
+    }
```
Comment (yannickperrenet) [inline services/orchest-webserver/app/app/core/jobs.py:164]: We should probably preface this with a `TODO` comment for once we move to Python3.9, since that includes https://www.python.org/dev/peps/pep-0584.

File: services/orchest-webserver/app/app/views/orchest_api.py (hunk 2 above)
Comment (ricklamers) [reply services/orchest-webserver/app/app/views/orchest_api.py:195]: Not sure if the information is present in that context. As those names can change maybe we can resolve the names on the client (to be the actuals) instead of having a layer of "materialized names/cached names".

File: services/orchest-webserver/app/app/core/jobs.py (hunk 5 above)
Comment (ricklamers) [reply services/orchest-webserver/app/app/core/jobs.py:164]: Although the new syntax can be considered an improvement. I don't think a TODO is warranted. Thanks for keeping up recent on the Python language @yannickperrenet!

**Comment by ricklamers** [comment]
LGTM! I think it's good if we can exhaustively test the PR for edge cases (it seems quite tricky to test). I think we can add it to Cypress to add basic coverage of the high level UI actions that you can take. As a first suggestion:

Duplicating a job where there are changes between the old and new parameters (cases: parameter newly added, already exists but different default, parameter removed).

So far I couldn't find any issues. Great job :fire:!

The e2e Cypress tests can be added later (can be a separate PR imo).

File: services/orchest-webserver/app/app/core/jobs.py (hunk 5 above)
Comment (ricklamers) [reply services/orchest-webserver/app/app/core/jobs.py:164]: P.S. Slightly out of scope I think for this PR.

File: services/orchest-webserver/app/app/views/orchest_api.py (hunk 2 above)
Comment (ricklamers) [reply services/orchest-webserver/app/app/views/orchest_api.py:195]: Slightly out of scope I think for this PR.

File: services/orchest-webserver/client/src/views/JobView.tsx (hunk 1 above)
Comment (ricklamers) [reply services/orchest-webserver/client/src/views/JobView.tsx]: Added this in a commit.

**Comment by ricklamers** [comment]
My last commit solves #432 by the way.

File: services/orchest-webserver/client/src/views/JobView.tsx [hunk 6]
```diff
@@ -389,6 +395,64 @@ const JobView: React.FC<TViewProps> = (props) => {
...
+              } catch (error) {
+                console.log(error);
+              }
+            }
+          });
+      })
+      .catch((result) => {
+        if (result.reason === "gate-failed") {
+          setTimeout(() => {
+            orchest.requestBuild(
```
Comment (ricklamers) [inline services/orchest-webserver/client/src/views/JobView.tsx:443]: Not sure why this requires the timeout.

Comment (fruttasecca) [reply services/orchest-webserver/client/src/views/JobView.tsx]: That's just me misinterpreting the existing FE logic :D.

File: services/orchest-webserver/app/app/core/jobs.py (hunk 3 above)
Comment (fruttasecca) [reply services/orchest-webserver/app/app/core/jobs.py:131]: Indeed, just a layer of "safety" for very improbable race conditions.

File: services/orchest-webserver/app/app/views/orchest_api.py (hunk 2 above)
Comment (fruttasecca) [reply services/orchest-webserver/app/app/views/orchest_api.py:195]: >Wouldn't it be more helpful if we add the environment names to the message as well?

I agree. I think this ties (once again) into #391. Let's keep this use case/issue in mind.

**Comment by fruttasecca** [comment]
@ricklamers 

About the tests, what we really want to test imo is `duplicate_job_spec` in `services/orchest-webserver/app/app/core/jobs.py `. It might be less costly to do that through the non-integration tests. Although the integration tests will be "better" since they are testing the real thing.

>The e2e Cypress tests can be added later (can be a separate PR imo).

I agree with that. If we are ok with adding some non-integration tests to this PR those might take a fraction of the time and allow us to sleep better.

**Comment by yannickperrenet** [comment]
> If we are ok with adding some non-integration tests to this PR

I like this idea. The fact that we are about to introduce integration tests doesn't mean we shouldn't ever write unit tests any more. And this test could indeed be a well scoped unit test.

File: services/orchest-webserver/app/app/core/jobs.py (hunk 5 above)
Comment (yannickperrenet) [reply services/orchest-webserver/app/app/core/jobs.py:164]: Works for me.

File: services/orchest-webserver/app/tests/test_jobs.py [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+
+def _get_parent_job_spec(
+    project_uuid="project-uuid",
+    pipeline_uuid="pipeline-uuid",
+    name="job-name",
+    schedule=None,
+    pipeline_name="pipeline-name",
+    env_variables={},
+    strategy_json={},
+    parameters=[{}],
```
Comment (yannickperrenet) [inline services/orchest-webserver/app/tests/test_jobs.py:21]: In Python you should not create mutable default arguments ([some explanation](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments) together with "correct" alternative).

Comment (fruttasecca) [reply services/orchest-webserver/app/tests/test_jobs.py:21]: Yep, my bad, will fix.

**Review by yannickperrenet — APPROVED** [review]
